src/executables/util_cs.c

```diff
@@ -3994,6 +3994,798 @@ tde (UTIL_FUNCTION_ARG * arg)
+	}
+
+      attr = attr->order_link;
+    }
+
+  if (is_detail)
+    {
+      ptr = *loginfo;
+
+      original_sql = (char *) malloc (max_original_size);
+      if (original_sql == NULL)
+	{
+	  error = ER_OUT_OF_VIRTUAL_MEMORY;
+	  goto error;
+	}
+
+      sprintf (original_sql, "insert into %s values ( ", classname);
+
+      ptr = or_unpack_int (ptr, &num_change_col);
+      for (i = 0; i < num_change_col; i++)
+	{
+	  ptr = or_unpack_int (ptr, &change_index);
+	}
+
+      attr = class_->ordered_attributes;
+      for (i = 0; i < num_change_col; i++)
+	{
+	  error =
+	    check_sql_memory (&original_sql, strlen (original_sql) + strlen (attr->header.name) + 12,
+			      &max_original_size);
+	  if (error != NO_ERROR)
+	    {
+	      goto error;
+	    }
+
+	  process_column_data (&ptr, &original_sql, &max_original_size, attr->type->id);
+	  if (i != num_change_col - 1)
+	    {
+	      strcat (original_sql, ", ");
+	    }
+	  else
+	    {
+	      strcat (original_sql, " )");
+	    }
+
+	  attr = attr->order_link;
+	}
+    }
+
+  ptr = or_unpack_int (ptr, &num_cond_col);
+
+  *loginfo = ptr;
+
+  if (is_detail)
+    {
+      print_flashback_detail (trid, user, sql, original_sql, outfp);
+    }
+  else
+    {
+      fprintf (outfp, "%s\n", sql);
+    }
+
+  free_and_init (sql);
+  if (original_sql != NULL)
+    {
+      free_and_init (original_sql);
+    }
+
+  return NO_ERROR;
+
+error:
+  if (sql != NULL)
+    {
+      free_and_init (sql);
+    }
+
+  if (original_sql != NULL)
+    {
+      free_and_init (original_sql);
+    }
+
+  return error;
+}
+
+static int
+print_flashback_info (char *loginfo, int num_item, dynamic_array * classlist, OID * oidlist, bool is_detail,
+		      FILE * outfp)
+{
+  int length;
+  TRANID trid;
+  char *user;
+  int dataitem_type;
+
+  int error = NO_ERROR;
+
+  int dml_type;
+  INT64 b_classoid;
+  OID classoid;
+  int class_index = 0;
+  char classname[SM_MAX_IDENTIFIER_LENGTH] = "\0";
+
+  char *ptr = loginfo;
+
+  for (int i = 0; i < num_item; i++)
+    {
+      ptr = PTR_ALIGN (ptr, MAX_ALIGNMENT);
+      ptr = or_unpack_int (ptr, &length);
+      ptr = or_unpack_int (ptr, &trid);
+      ptr = or_unpack_string_nocopy (ptr, &user);
+      ptr = or_unpack_int (ptr, &dataitem_type);
+
+      ptr = or_unpack_int (ptr, &dml_type);
+      ptr = or_unpack_int64 (ptr, &b_classoid);
+      memcpy (&classoid, &b_classoid, sizeof (OID));
+
+      class_index = get_class_index (da_size (classlist), oidlist, classoid);
+
+      da_get (classlist, class_index, classname);
+
+      switch (dml_type)
+	{
+	case 0:
+	  /* INSERT */
+	  error = print_flashback_insert (&ptr, trid, user, classname, is_detail, outfp);
+	  break;
+	case 1:
+	  /* UPDATE */
+	  error = print_flashback_update (&ptr, trid, user, classname, is_detail, outfp);
+	  break;
+	case 2:
+	  /* DELETE */
+	  error = print_flashback_delete (&ptr, trid, user, classname, is_detail, outfp);
+	  break;
+	default:
+	  assert (dml_type < 3 || dml_type >= 0);
```

Comment:
Maybe `dml_type < 3 && dml_type >= 0`? and it never happens.
```suggestion
	  assert (false);
```

Reply:
OK, I will fix it.